### PR TITLE
Python HTML report generator

### PR DESCRIPTION
**Pull request by colinjennings**
@colinjennings , which of these items have been done since our meeting yesterday?

- [x] Move Colin's files to his own folder
- [x] Upload Colin's code to his personal GitHub and submit a pull request to this GitHub
- [x] Add percentage labels on pie chart
- [x] Remove labels and strings off pies that are <10 %
- [x] At the bottom of the html report, have hyperlinked granularity of folders of the biggest 5 parent folders
- [x] After a hyperlink is clicked, move focus to the rendered table so the user do not wonder about if anything has happened (optional, can leave if too hard)
- [x] Tashrif will fix mailreport.sh
- [x] Remove the HH:mm timestamp from wherever needed in the html report (must remove `:SS` from all timestamps)
- [x] Next day, Colin will demo how his program works when there are multiple dirs in `_config/dirs.txt`
- [x] Add a variable month/year to look from, default should be last 12 months
- [x] Follow color scheme of `report-20231224.html` as reference
- [x] Change to only month/day such as 5/18 - 5/25:
  > 1 Week Changes for /data/pnl/ (2024-05-18 - 2024-05-25)
- [x] Round all GB sizes to two decimal points only
- [x] Can you make the percentage label transparent or smaller? As you see, one is eclipsing the other now.
- [x] Can you set it up in a way that after the expanded directory is clicked again, the table disappears from the view?

### CODE REVIEW

**Review by RyanZurrin — APPROVED** [review]

**Comment by tashrifbillah** [comment]
@colinjennings , which of these items have been done since our meeting yesterday?

- [x] Move Colin's files to his own folder
- [ ] Upload Colin's code to his personal GitHub and submit a pull request to this GitHub
- [x] Add percentage labels on pie chart
- [ ] Remove labels and strings off pies that are <10 %
- [x] At the bottom of the html report, have hyperlinked granularity of folders of the biggest 5 parent folders
- [ ] After a hyperlink is clicked, move focus to the rendered table so the user do not wonder about if anything has happened (optional, can leave if too hard)
- [x] Tashrif will fix mailreport.sh
- [ ] Remove the HH:mm timestamp from wherever needed in the html report (must remove `:SS` from all timestamps)
- [x] Next day, Colin will demo how his program works when there are multiple dirs in `_config/dirs.txt`
- [ ] Add a variable month/year to look from, default should be last 12 months
- [ ] Follow color scheme of `report-20231224.html` as reference
- [ ] Change to only month/day such as 5/18 - 5/25:
  > 1 Week Changes for /data/pnl/ (2024-05-18 - 2024-05-25)
- [ ] Round all GB sizes to two decimal points only

**Comment by tashrifbillah** [comment]
Primary problem of this work is that `matlplotlib` is not properly set up. Hence, the feature is not working without the availability of GUI. It worked last time because we were sitting at our computer. But `cron` will not have any GUI:

```
**PuTTY X11 proxy: unable to connect to forwarded X server: Network error: Connection refused
qt.qpa.xcb: could not connect to display localhost:10.0
qt.qpa.plugin: Could not load the Qt platform plugin "xcb" in "" even though it was found.
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

Available platform plugins are: eglfs, minimal, minimalegl, offscreen, vnc, webgl, xcb.

Aborted (core dumped)
```

---

Edit:
Fixed by: https://github.com/pnlbwh/diskusage-logging/commit/ab62e5391ad5c35b9c6b1fae59e58c85603e1e30

**Comment by tashrifbillah** [comment]
You can follow my code to fix the issue:

https://github.com/pnlbwh/TBSS/blob/2fb1bc776b07013b450c24057db876486fc4bb2b/lib/writeHtml.py#L23-L26

---

Edit:
Fixed by: https://github.com/pnlbwh/diskusage-logging/commit/ab62e5391ad5c35b9c6b1fae59e58c85603e1e30

File: html-report-generator/generate_usage_report.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
+#!/usr/bin/env python
```
Comment (tashrifbillah) [inline html-report-generator/generate_usage_report.py:1]: Run `chmod a+x generate_usage_report.py` and then push this file up.

**Comment by tashrifbillah** [comment]
TODO for Tashrif and Colin--
- [x] Update the `report-${datestamp}.html` name to `report-rfanfs-${datestamp}.html` and `report-data-${datestamp}.html`. 
- [x] Also update the subject in https://github.com/pnlbwh/diskusage-logging/blob/17596017c4da58bc29ba5e922d629cf28ae4e98e/mailreport.sh#L17

File: html-report-generator/generate_usage_report.py [hunk 2]
```diff
@@ -36,6 +33,13 @@ def main():
     with open(config_file_path, 'r') as file:
         dir_list = [line.strip() for line in file]
 
+
+    # Get the name of the filesystem from config.txt
+    filesystem_name = dir_list[0].split('/')[1]
+
+    # Title report file as report-data-{datestamp}.html or report-rfanfs-{datestamp}.html
+    file_path = f"{pjoin(working_dir, '_data/', 'htmlreport/')}report-{filesystem_name}-{current_date:%Y%m%d}.html"
```
Comment (tashrifbillah) [inline html-report-generator/generate_usage_report.py]: Did you test it? `()` within `f"{ }"` do not usually work. Hence, we had to stick to `.format()`

File: html-report-generator/generate_usage_report.py [hunk 3]
```diff
@@ -1,33 +1,30 @@
 #!/usr/bin/env python
 
 import datetime
-import sys
 import itertools
 from os.path import dirname, abspath, basename, join as pjoin
 
+# Use Agg non-interactive backend to prevent X forwarding error
+import matplotlib
+matplotlib.use('Agg')
```
Comment (tashrifbillah) [inline html-report-generator/generate_usage_report.py:13]: Why is this coming up in diff? I had already committed it to the PR that you merged.

Comment (tashrifbillah) [reply html-report-generator/generate_usage_report.py:13]: Did you shuffle around the sequence of imports within this script?

**Comment by tashrifbillah** [comment]
@colinjennings , check off items that are done:
https://github.com/pnlbwh/diskusage-logging/pull/13#issuecomment-2142930904

**Comment by colinjennings** [comment]
> @colinjennings , check off items that are done: [#13 (comment)](https://github.com/pnlbwh/diskusage-logging/pull/13#issuecomment-2142930904)

Unable to check things off the list but I have completed all of them.

File: html-report-generator/generate_usage_report.py (hunk 1 above)
Comment (tashrifbillah) [reply html-report-generator/generate_usage_report.py:1]: Tashrif did it.

**Comment by tashrifbillah** [comment]
Sorry, I did not realize your lack of permission. Can you check things off at the top of this page now?

**Comment by tashrifbillah** [comment]
Also, they way GitHub PR collaboration works is that you need to go through unresolved conversations and put your replies.

**Comment by tashrifbillah** [comment]
Also, the way GitHub PR collaboration works is that you need to go through unresolved conversations and put your replies.

File: html-report-generator/generate_usage_report.py (hunk 2 above)
Comment (colinjennings) [reply html-report-generator/generate_usage_report.py]: Tested with /rfanfs/pnl-zorro/ and /data/pnl/ and both worked, could convert to format if wanted.

File: html-report-generator/generate_usage_report.py (hunk 3 above)
Comment (colinjennings) [reply html-report-generator/generate_usage_report.py:13]: Moved it above the other imports, as we want to initialize matplotlib with Agg before the other scripts import pyplot in the other files.

**Comment by tashrifbillah** [comment]
Thank you for addressing the comments so far. I am waiting for weekend's `/data/` run to complete. `/rfanfs/` run completed successfully though.

**Comment by tashrifbillah** [comment]
Colin, `/data/` report just came out. Here are the issues we noticed:

image


1. Can you make the percentage label transparent or smaller? As you see, one is eclipsing the other now.

2. Can you set it up in a way that after the expanded directory is clicked again, the table disappears from the view?

image

**Review by tashrifbillah — COMMENTED** [review]
`/rfanfs/` and `/data/` html reports completed successfully over the weekend. Thanks to Colin for his contribution. We shall do a few minor tweaks before merging it to master.

**Comment by colinjennings** [comment]
Would it be acceptable to nudge the bottom label down if it is going to collide with the top label?

**Comment by tashrifbillah** [comment]
How would you know when it is going to collide upward or downward?

**Comment by tashrifbillah** [comment]
Colin, please put the latest graph screenshot here for record. The one that you sent me in Teams.

**Comment by colinjennings** [comment]
Screenshot 2024-06-05 at 12 09 55 PM

**Comment by colinjennings** [comment]
- [x] Move size increase column
~~- [x] Convert to percent~~
- [x] Only display directory changes 3 layers deep
- [x] Do not show things with no size increase

**Comment by tashrifbillah** [comment]
Colin, I have deployed your latest commit. So it is going through another round of testing this weekend. Also, update the above checklist if appropriate. I feel like you completed the last item too.

File: _config/dirs.txt [hunk 4]
```diff
@@ -1,2 +1,3 @@
-/rfanfs/pnl-zorro/
-tb571@eris1n3.research.partners.org:/data/pnl
+/data/pnl/
+/data/pnlx/
+/data/predict1/
```
Comment (tashrifbillah) [inline _config/dirs.txt]: Colin, I want you to undo this commit. It was right before. This `dirs.txt` serves as an example and so we want it to contain both rfanfs and data.

File: html-report-generator/generate_usage_report.py [hunk 5]
```diff
@@ -31,7 +31,8 @@ def main():
     # Get directories to query from _config/dirs.txt
     config_file_path = pjoin(working_dir, '_config/') + 'dirs.txt'
     with open(config_file_path, 'r') as file:
-        dir_list = [line.strip() for line in file]
```
Comment (tashrifbillah) [inline html-report-generator/generate_usage_report.py]: The way we experts do it is that:

> dir_list = file.read().strip().split('\n')

The last can be left as just `.split()` if you are confident that no line has whitespace in it.

**Comment by tashrifbillah** [comment]
Either way, I think you can undo the two commits you made today morning.

File: html-report-generator/generate_usage_report.py (hunk 5 above)
Comment (colinjennings) [reply html-report-generator/generate_usage_report.py]: Tested and fixed

File: _config/dirs.txt (hunk 4 above)
Comment (colinjennings) [reply _config/dirs.txt]: I undid the commit, but I am keeping dirs.txt in .gitignore to prevent issues when testing.

**Comment by tashrifbillah** [comment]
I confirm that the Friday evening's commit solved the issue of showing parent and children directories under change sections. 
Now the directories are unique.

image

File: html-report-generator/generate_usage_report.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+        return output_html
+
+
+    # Define the HTML for the report
+    html_content = f"""
+    <!DOCTYPE html>
+    <html>
+    <head>
+        <meta charset="UTF-8">
+        <title>Disk Report</title>
```
Comment (tashrifbillah) [inline html-report-generator/generate_usage_report.py]: Change this to `Disk Usage Report`.

File: _config/dirs.txt [hunk 7]
```diff
@@ -1,2 +0,0 @@
-/rfanfs/pnl-zorro/
-tb571@eris1n3.research.partners.org:/data/pnl
```
Comment (tashrifbillah) [inline _config/dirs.txt]: @colinjennings , this is still incorrect. Bring back our original `_config/dirs.txt`. I see that you added it to `.gitignore` already. So its changes will get out of your way in future.

Comment (colinjennings) [reply _config/dirs.txt]: Reverted to original state and used `git update-index --assume-unchanged _config/dirs.txt` to make sure the contents are not tracked

**Comment by colinjennings** [comment]
- [x] Sentence case where applicable
- [x] Left-shift titles
- [x] Make sure full directories are spelled out 
- [x] Align tables

**Comment by tashrifbillah** [comment]
New TODOs:

- [ ] solve the bug at https://github.com/colinjennings/diskusage-logging/blob/d0049b70c7b5873a008a792d9c0e02e8e487b0be/html-report-generator/get_biggest_changes.py#L33
- [ ] add a requirements.txt at root level with three packages: jinja2, matplotlib, pandas
- [ ] revise installation instruction with something like `pip install -r requirements.txt` instead of all the R stuff. Colin can have liberty in this regard.

**Comment by colinjennings** [comment]
could you please clarify the nature of the bug at https://github.com/colinjennings/diskusage-logging/blob/d0049b70c7b5873a008a792d9c0e02e8e487b0be/html-report-generator/get_biggest_changes.py#L33

**Comment by tashrifbillah** [comment]
Colin, we can wait until your access to email is restored. I wrote everything in email.

**Comment by tashrifbillah** [comment]
Can you message tbillah [at] bwh.harvard.edu from your personal email if comfortable to do so?

**Comment by tashrifbillah** [comment]
Colin, this is the checklist you need to work on here or privately on every commit:

```
- [ ] Pushed up commit
- [ ] Downloaded code to /tmp/
- [ ] copy /to/official/location
- [ ] Run generate_usage_report.py and confirm success
- [ ] Repeat steps 2-4 in both rfanfs and data
```

**Comment by colinjennings** [comment]
- [x] Pushed up commit
- [x] Downloaded code to /tmp/
- [x] copy /to/official/location
- [x] Run generate_usage_report.py and confirm success
- [x] Repeat steps 2-4 in both rfanfs and data
- [x] Update installation instructions
- [x] Remove time string from title

**Comment by tashrifbillah** [comment]
image

Hi Colin, can you remove the `05:57` part? This info is redundant for the scope of this report.

**Comment by tashrifbillah** [comment]
Colin, it appears that you did not deploy this change. Just check the report you got in email and you will see that the time string still exists.

**Comment by colinjennings** [comment]
> Colin, it appears that you did not deploy this change. Just check the report you got in email and you will see that the time string still exists.

My mistake Tashrif -- just deployed the changes.

**Comment by tashrifbillah** [comment]
Thank you Colin. The report looks good now. I plan to find a time to sit with you and complete code review. We have reviewed output so far.

Also, I would like to review the installation instruction with you.

**Comment by tashrifbillah** [comment]
Hi Colin, I would like to merge your PR today.

**Pull request merged**